File: lib/vbrain/page.rb
require "yaml"
require "digest"
require "date"
require "fileutils"

module VBrain
  module Page
    FRONTMATTER_RE = /\A---\s*\n(.*?)\n---\s*\n(.*)\z/m

    class Error < StandardError; end

    Parsed = Struct.new(:frontmatter, :body, :sha256, keyword_init: true)

    def self.parse(path)
      content = File.read(path)
      parse_string(content)
    end

    def self.parse_string(content)
      if (m = content.match(FRONTMATTER_RE))
        fm = YAML.safe_load(m[1], permitted_classes: [Symbol, Date, Time]) || {}
        body = m[2]
      else
        fm = {}
        body = content
      end
      Parsed.new(frontmatter: fm, body: body, sha256: Digest::SHA256.hexdigest(body))
    end

    def self.write(dir:, slug:, frontmatter:, body:)
      raise Error, "dir must exist" unless Dir.exist?(dir)
      raise Error, "slug cannot be empty" if slug.nil? || slug.empty?

      full = File.join(dir, "#{slug}.md")
      content = render(frontmatter, body)

      tmp = "#{full}.tmp.#{Process.pid}.#{rand(1 << 32)}"
      File.write(tmp, content)
      File.rename(tmp, full)

      full
    end

    # Reescreve só o CORPO de uma página, preservando o frontmatter verbatim
    # (sem reserializar YAML — zero churn de formatação/ordem de chaves).
    # Passa o body atual ao bloco; grava de volta atomicamente só se mudou.
    # Retorna true se reescreveu, false se o bloco devolveu o body inalterado.
    def self.rewrite_body!(path)
      content = File.read(path)
      m = content.match(FRONTMATTER_RE)
      body = m ? m[2] : content
      new_body = yield(body)
      return false if new_body == body

      new_content = m ? "---\n#{m[1]}\n---\n#{new_body}" : new_body
      tmp = "#{path}.tmp.#{Process.pid}.#{rand(1 << 32)}"
      File.write(tmp, new_content)
      File.rename(tmp, path)
      true
    end

    def self.render(frontmatter, body)
      "#{YAML.dump(stringify_keys(frontmatter))}---\n#{body}"
    end

    def self.stringify_keys(hash)
      hash.each_with_object({}) do |(k, v), out|
        out[k.to_s] = v
      end
    end
  end
end

File: internal/page/page.go
// Package page lê e escreve páginas markdown com frontmatter YAML. Porta
// determinística de lib/vbrain/page.rb.
package page

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"regexp"

	"gopkg.in/yaml.v3"
)

// frontmatterRE separa o bloco YAML do corpo. (?s) = `.` casa newline (flag /m
// do Ruby).
var frontmatterRE = regexp.MustCompile(`(?s)\A---\s*\n(.*?)\n---\s*\n(.*)\z`)

// ErrInvalid cobre dir inexistente ou slug vazio no Write.
var ErrInvalid = errors.New("page: dir must exist and slug cannot be empty")

// Parsed é o resultado de parse: frontmatter, corpo e sha256 do corpo.
type Parsed struct {
	Frontmatter map[string]any
	Body        string
	SHA256      string
}

// Parse lê e parseia uma página do disco.
func Parse(path string) (Parsed, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return Parsed{}, err
	}
	return ParseString(string(content))
}

// ParseString parseia conteúdo em memória. Sem frontmatter, body = conteúdo
// inteiro e frontmatter vazio.
func ParseString(content string) (Parsed, error) {
	fm := map[string]any{}
	body := content
	if m := frontmatterRE.FindStringSubmatch(content); m != nil {
		if err := yaml.Unmarshal([]byte(m[1]), &fm); err != nil {
			return Parsed{}, err
		}
		if fm == nil {
			fm = map[string]any{}
		}
		body = m[2]
	}
	return Parsed{Frontmatter: fm, Body: body, SHA256: sha256hex(body)}, nil
}

// Write renderiza frontmatter+body e grava atomicamente em dir/<slug>.md.
func Write(dir, slug string, frontmatter map[string]any, body string) (string, error) {
	if fi, err := os.Stat(dir); err != nil || !fi.IsDir() {
		return "", ErrInvalid
	}
	if slug == "" {
		return "", ErrInvalid
	}

	full := filepath.Join(dir, slug+".md")
	content, err := Render(frontmatter, body)
	if err != nil {
		return "", err
	}
	if err := atomicWrite(full, content); err != nil {
		return "", err
	}
	return full, nil
}

// RewriteBody reescreve só o CORPO, preservando o frontmatter verbatim (sem
// reserializar YAML — zero churn). Retorna true se gravou, false se inalterado.
func RewriteBody(path string, transform func(string) string) (bool, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return false, err
	}
	content := string(raw)

	m := frontmatterRE.FindStringSubmatch(content)
	body := content
	if m != nil {
		body = m[2]
	}
	newBody := transform(body)
	if newBody == body {
		return false, nil
	}

	newContent := newBody
	if m != nil {
		newContent = "---\n" + m[1] + "\n---\n" + newBody
	}
	if err := atomicWrite(path, newContent); err != nil {
		return false, err
	}
	return true, nil
}

// Render serializa frontmatter (chaves string) + corpo no formato da wiki.
func Render(frontmatter map[string]any, body string) (string, error) {
	out, err := yaml.Marshal(frontmatter)
	if err != nil {
		return "", err
	}
	return "---\n" + string(out) + "---\n" + body, nil
}

func atomicWrite(full, content string) error {
	tmp := fmt.Sprintf("%s.tmp.%d.%d", full, os.Getpid(), rand.Uint32())
	if err := os.WriteFile(tmp, []byte(content), 0o644); err != nil {
		return err
	}
	return os.Rename(tmp, full)
}

func sha256hex(s string) string {
	sum := sha256.Sum256([]byte(s))
	return hex.EncodeToString(sum[:])
}
